reps/2023-03-20-air-storage-path.md

```diff
@@ -0,0 +1,248 @@
+in the configuration.
+
+Risk: Since we will continue to use locally cached data for the implementation,
+hiding the location in e.g. an environment variable can make failures more opaque
+to users.
+
+
+### Should this change be within `ray` or outside?
+
+This is an API change to Ray AIR (specifically, `air.RunConfig`, `tune.SyncConfig`, and `tune.run()`)
+and therefore has to be within Ray.
+
+
+## Stewardship
+
+### Required Reviewers
+
+@richardliaw, @ericl, @gjoliver
+
+### Shepherd of the Proposal (should be a senior committer)
+
+@richardliaw
+
+## Design and Architecture
+
+- We introduce a new `storage_path` argument to `air.RunConfig` and `tune.run` as the main configuration entrypoints
+- The `storage_path` argument defaults to the configured `RAY_STORAGE` environment variable
+- If the `storage_path` is set to a remote URI, the `local_path` is read from an environment variable `RAY_AIR_CACHE_DIR`
+- Backwards compatibility: If a `local_dir` is passed, we set this environment variable
+- In downstream components (`TrialRunner`, `Experiment`, `Trial`), we introduce respective arguments: `storage_path` and `experiment_path`
+
+We retain full backwards compatibility with the current API.
+
+## Synchronization
+
+The synchronization to a common location already works today. However, we can improve this
+for better efficiency and to better define the responsibilities of different components.
+
+### Current implementation
+
+The current flow multiplexes between two scenarios: With cloud storage enabled or without it.
+
+When cloud storage is enabled:
+
+- The driver saves experiment data to `/local/path/experiment_name`
+- The driver saves driver-based trial data to `/local/path/experiment_name/trial_id`
+- The driver uses a `Syncer` to periodically sync this state to the remote storage (e.g. S3), e.g. `s3://bucket/location/experiment_name`
+- This _includes_ driver-based trial data, such as the config and results log
+- This _excludes_ trainable-based trial data, such as checkpoints.
+  Uploading checkpoints is the trainable's responsibility.
+- The trainable saves trial-based data such as checkpoints to their local `/local/path/experiment_name/trial_id`
+- The trainable uses a `Syncer` to upload these checkpoints _synchronously_ to the remote storage after saving, e.g. `s3://bucket/location/experiment_name/trial_id/checkpoint_001`
+
+When cloud storage is disabled:
+
+- The driver saves experiment data to `/local/path/experiment_name`
+- The driver saves driver-based trial data to `/local/path/experiment_name/trial_id`
+- The trainable saves trial-based data such as checkpoints to their local `/local/path/experiment_name/trial_id`
+- The driver uses a `SyncerCallback` to synchronize contents form the trial directories (i.e. the missing trial-based data).
+  This uses the Ray object store.
+- This also happens _synchronously_ every time a trial reports a checkpoint has been saved
```

Comment:
What are the limitations of this mechanism? Is there some max size limit?

Reply:
Good point. The data must fit into the heap memory. I've added a comment.